### PR TITLE
Add GST1 B2B reports

### PR DESCRIPTION
**Pull request by ankitsinghaniyaz**

### CODE REVIEW

File: src/utils.js [hunk 1]
```diff
@@ -302,3 +306,20 @@ export function fuzzyMatch(keyword, candidate) {
 export function openSettings(tab) {
   routeTo({ path: '/settings', query: { tab } });
 }
+
+export function promptWhenGstUnavailable() {
```
Comment (ankitsinghaniyaz) [inline src/utils.js:310]: This function need not be a part of utils, it can move inside gst.js

File: src/saveReportAsJson.js [hunk 2]
```diff
@@ -0,0 +1,9 @@
+import fs from 'fs';
+import { shell } from 'electron';
+
+export default async function makeJSON(data, savePath) {
+  fs.writeFile(savePath, data, (error) => {
+    if (error) throw error;
```
Comment (ankitsinghaniyaz) [inline src/saveReportAsJson.js:6]: Should we show a modal to user in case the json fails?

File: src/sidebarConfig.js [hunk 3]
```diff
@@ -103,6 +103,14 @@ const config = {
...
           route: '/report/trial-balance',
         },
+        {
+          label: _('GSTR1'),
+          route: '/report/gstr-1',
+        },
+        {
+          label: _('GSTR2'),
+          route: '/report/gstr-2',
+        },
```
Comment (ankitsinghaniyaz) [inline src/sidebarConfig.js:113]: These routes should be conditional and should only shop up when country `India` is selected.

File: src/pages/Report.vue [hunk 4]
```diff
@@ -304,8 +313,11 @@ export default {
           return `minmax(${minWidth}, ${maxWidth})`;
         })
         .join(' ');
-    }
-  }
+    },
+    isGstReportsPage() {
```
Comment (ankitsinghaniyaz) [inline src/pages/Report.vue:317]: this is not required anymore

File: reports/GoodsAndServiceTax/GSTR1View.js [hunk 5]
```diff
@@ -1,10 +1,22 @@
...
 import baseConfig from './BaseViewConfig';
+import { DateTime } from 'luxon';
+
+const filterFields = baseConfig.filterFields;
+filterFields.forEach((field) => {
+  if (field.fieldname === 'toDate') {
+    field.default = DateTime.local().toISODate();
+  } 
+  else if(field.fieldname === 'fromDate') {
+    field.default = DateTime.local().minus({months: 3}).toISODate();
```
Comment (ankitsinghaniyaz) [inline reports/GoodsAndServiceTax/GSTR1View.js:11]: I think these filters are not working correctly?

In the case of `toDate` and `fromDate` the same `default` field is set. can you please verify once?

File: reports/GoodsAndServiceTax/GSTR1.js [hunk 6]
```diff
@@ -14,10 +14,10 @@ class GSTR1 extends BaseGSTR {
     const data = await this.getCompleteReport('GSTR-1', filters);
 
     const conditions = {
-      B2B: row => row.gstin,
+      'B2B': row => row.gstin,
       'B2C-Large': row => !row.gstin && !row.inState && row.invAmt >= 250000,
-      'B2C-Small': row =>
-        !row.gstin && (row.inState || (row.inState && row.invAmt < 250000))
+      'B2C-Small': row => !row.gstin && (row.inState || (row.inState && row.invAmt < 250000)),
+      'Nil Rated, Exempted and Non GST supplies': row => (row.rate === 0), // this takes care of both nil rated, exempted goods
```
Comment (ankitsinghaniyaz) [inline reports/GoodsAndServiceTax/GSTR1.js:20]: :+1:

File: src/sidebarConfig.js (hunk 3 above)
Comment (piyushsinghania) [reply src/sidebarConfig.js:113]: Good point @ankitsinghaniyaz, I'll do the change.

File: reports/GoodsAndServiceTax/GSTR1View.js (hunk 5 above)
Comment (piyushsinghania) [reply reports/GoodsAndServiceTax/GSTR1View.js:11]: image
Here in my machine this is what I get @ankitsinghaniyaz

File: src/pages/Report.vue (hunk 4 above)
Comment (piyushsinghania) [reply src/pages/Report.vue]: Done 😄 (deleted)

File: src/utils.js (hunk 1 above)
Comment (piyushsinghania) [reply src/utils.js]: Done 💯

File: reports/GoodsAndServiceTax/GSTR1View.js (hunk 5 above)
Comment (piyushsinghania) [reply reports/GoodsAndServiceTax/GSTR1View.js:11]: May be in the GST R2 page it might not be correct, so updating there as well

Comment (piyushsinghania) [reply reports/GoodsAndServiceTax/GSTR1View.js:11]: Found a better way , I have updated the filters 💯

File: src/sidebarConfig.js (hunk 3 above)
Comment (piyushsinghania) [reply src/sidebarConfig.js:113]: Done 💯

File: src/saveReportAsJson.js (hunk 2 above)
Comment (piyushsinghania) [reply src/saveReportAsJson.js:6]: @ankitsinghaniyaz I assume it means writeFile fails (and an error occurs)?

File: accounting/gst.js [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+  'IGST-12': 12,
+  'IGST-18': 18,
+  'IGST-28': 28,
+};
+
+export async function generateGstr1Json(report, { transferType, toDate }) {
+  const printSettings = await frappe.getSingle('PrintSettings');
+
+  if (!printSettings.gstin) promptWhenGstUnavailable();
+  else {
```
Comment (ankitsinghaniyaz) [inline accounting/gst.js:66]: No need to use else block, it add an extra level of indentatio in the app. We can rather reply on using if and return like

```suggestion
  if (!printSetting.gsting) {
    prompt();
    return
  }
```

File: accounting/gst.js [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+      gstData.b2cs = await getB2csData(report.rows);
+    }
+
+    await sleep(1);
+    const jsonData = JSON.stringify(gstData);
+    makeJSON(jsonData, savePath);
+  }
+}
+
+async function getB2bData(invoices) {
```
Comment (ankitsinghaniyaz) [inline accounting/gst.js:95]: let's rename these methods to use `generate` instead of word `get`
example: `generateB2bData`

Comment (ankitsinghaniyaz) [reply accounting/gst.js:95]: We should do the same for all types

File: src/components/Sidebar.vue [hunk 9]
```diff
@@ -124,6 +125,16 @@ export default {
...
 
+    if (country !== 'India') {
+      this.groups.forEach((group) => {
+        if (group.title === 'Reports') {
+          group.items = group.items.filter(
+            (item) => !item.label.toLowerCase().includes('gst')
+          );
+        }
+      });
+    }
```
Comment (ankitsinghaniyaz) [inline src/components/Sidebar.vue:136]: Is there a cleaner way to do this @piyushsinghania ?

File: src/saveReportAsJson.js (hunk 2 above)
Comment (ankitsinghaniyaz) [reply src/saveReportAsJson.js:6]: This is not a priority but could be a good thing. For example if the download file fails if the user is trying to save where he doesn't have write permission etc. An error could help the user understand why the download didn't happened and make it easy, but can be skipped for now.

File: accounting/gst.js (hunk 7 above)
Comment (piyushsinghania) [reply accounting/gst.js]: Done 💯

File: accounting/gst.js (hunk 8 above)
Comment (piyushsinghania) [reply accounting/gst.js]: Updated 💯

File: src/saveReportAsJson.js (hunk 2 above)
Comment (piyushsinghania) [reply src/saveReportAsJson.js:6]: Yes that would be a good thing, we will do this.

File: src/components/Sidebar.vue (hunk 9 above)
Comment (piyushsinghania) [reply src/components/Sidebar.vue]: @ankitsinghaniyaz done 💯 
Added ability to hide or show sidebar items,
this can be reused on adding further items.

Comment (piyushsinghania) [reply src/components/Sidebar.vue]: Done in a much simpler and cleaner way, thanks @ankitsinghaniyaz 💯

File: accounting/gst.js [hunk 10]
```diff
@@ -0,0 +1,197 @@
...
+
+export async function generateGstr1Json(report, { transferType, toDate }) {
+  const printSettings = await frappe.getSingle('PrintSettings');
+
+  if (!printSettings.gstin) {
+    promptWhenGstUnavailable();
+    return;
+  }
+
+  const savePath = await getSavePath();
```
Comment (ankitsinghaniyaz) [inline accounting/gst.js:70]: send the name from here 
```suggestion
  const savePath = await getSavePath('gstr-1');
```

**Pull request closed without merging**